### DISCUSSION TITLE
Malloc implementation comparison

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21936] [General]
Do we have any performance comparison between the emmalloc, dlmalloc, and mimalloc? I just know that mimalloc is fast, but heavyweight, while emmalloc is light, but I don't know the speed. dlmalloc is supposed to be somewhere between.

**Comment by kripken** [comment]
There are some numbers here for mimalloc vs dlmalloc:

https://github.com/emscripten-core/emscripten/pull/20651

Overall the bottom line is:

* emmalloc is tiny and often fast enough for applications that don't do heavy allocation.
* dlmalloc is a good single-threaded allocator. It is much larger than emmalloc, but it can be several times faster on specific benchmarks (e.g. Havlak in our benchmark suite). It is single-threaded, though, so heavily multithreaded applications may need more.
* mimalloc is even larger than dlmalloc because it also optimizes for multithreading. In some multithreaded cases it can be far faster than dlmalloc because it scales properly with more cores.
